File: Leetcode/2316 Count Unreachable Pairs of Nodes in an Undirected Graph.ts
function countPairs(n: number, edges: number[][]): number {
    const adjList: number[][] = new Array(n);
    for(let i = 0; i < n;i++) {
        adjList[i] = [];
    }
    for(const [src, dst] of edges) {
        adjList[src].push(dst);
        adjList[dst].push(src);
    }
    const visited: boolean[] = new Array(n).fill(false);
    let unreachablePairs = 0, remainingNodes = n;
    for(let node = 0; node < n; node++) {
        if(!visited[node]) {
            const componentSize = dfs(node, adjList, visited);
            unreachablePairs += componentSize * (remainingNodes - componentSize);
            remainingNodes -= componentSize;
        }
    }
    return unreachablePairs;
}

function dfs(start: number, adjList: number[][], visited: boolean[]) {
    visited[start] = true;
    let componentSize = 1;
    for (const neighbour of adjList[start]) {
        if(!visited[neighbour]) {
            componentSize += dfs(neighbour, adjList, visited);
        }
    }
    return componentSize;
}

File: Leetcode/2316 Count Unreachable Pairs of Nodes in an Undirected Graph.py
class Solution:
    def countPairs(self, n, edges):
        self.adj_list = defaultdict(list)
        for src, dst in edges:
            self.adj_list[src].append(dst)
            self.adj_list[dst].append(src)
        
        self.visited = [False] * n
        remaining_node, unreachable_pairs = n, 0
        for node in range(n):
            if not self.visited[node]:
                component_size = self.dfs(node)
                unreachable_pairs += component_size * (remaining_node - component_size)
                remaining_node -= component_size
        return unreachable_pairs
        
    def dfs(self, start):
        self.visited[start] = True
        component_size = 1
        for neighbour in self.adj_list[start]:
            if not self.visited[neighbour]:
                component_size += self.dfs(neighbour)
        return component_size
        
